src/idp_fraud_data_handler.py

```diff
@@ -0,0 +1,130 @@
+import csv
+import logging
+import os
+import re
+
+import dateparser
+
+from src.common import get_database_password
+from src.database import create_db_connection, write_import_session, write_idp_fraud_event_to_database, \
+    update_session_as_validated, write_upload_error
+from src.idp_fraud_event import IdpFraudEvent
+from src.s3 import fetch_object_tags, move_file, download_import_file
+from src.upload_session import UploadSession
+
+SUCCESS_FOLDER = 'success'
+ERROR_FOLDER = 'error'
+DEFAULT_TIMEZONE = 'Europe/London'
+DEFAULT_HAS_HEADER = True
+DEFAULT_DIALECT = 'excel'
+logger = logging.getLogger('idp_fraud_data_handler')
+logger.setLevel(logging.INFO)
+
+
+def create_import_session(filename, idp_entity_id, userid, db_connection):
+    upload_session = UploadSession(
+        source_file_name=filename,
+        idp_entity_id=idp_entity_id,
+        userid=userid
+    )
+    return write_import_session(upload_session, db_connection, logger)
+
+
+def process_file(bucket, filename, upload_session, db_connection,
+                 has_header=DEFAULT_HAS_HEADER, dialect=DEFAULT_DIALECT, timezone=DEFAULT_TIMEZONE):
+    logger.info('Processing data for IDP {}'.format(upload_session.idp_entity_id))
+
+    temp_file = download_import_file(bucket, filename)
+    with open(temp_file, newline='') as csvfile:
+        reader = csv.reader(csvfile, dialect=dialect)
+        errors_occurred = False
+        skip_row = has_header
+        row_number = 0
+
+        for row in reader:
+            row_number = row_number + 1
+            if skip_row:
+                skip_row = False
+                continue
+
+            try:
+                idp_fraud_event = parse_line(row, upload_session.idp_entity_id, timezone)
+                event_id = write_idp_fraud_event_to_database(upload_session, idp_fraud_event, db_connection, logger)
+                if event_id:
+                    logger.info(
+                        'Successfully wrote IDP fraud event ID {} to database and found matching fraud event {}'.format(
+                            idp_fraud_event.idp_event_id, event_id
+                        )
+                    )
+                else:
+                    logger.warning(
+                        'Successfully wrote IDP fraud event ID {} to database BUT no matching fraud event found'.format(
+                            idp_fraud_event.idp_event_id
+                        )
+                    )
+
+            except Exception as exception:
```

Comment:
Is catching all exceptions the right thing to do _inside the loop_? Can we be more specific about which exceptions we might want to ignore and carry on processing the file, and which ones might mean just giving up altogether and reporting failure sooner rather than later?

Reply:
I've re-factored this anyway, due to our conversations about rollbacking any successful `INSERT`s if errors are detected. This PR was only ever to concentrate on success cases, so having the catch-all outside the loop is fine, we can add finer grained validation and exception handling within the loop at a later date.